### PR TITLE
KAFKA-7204: Avoid clearing records for paused partitions on poll of MockConsumer

### PR DESCRIPTION
**Pull request by efgpinto**
The previous version of MockConsumer does not allow the clients to test consecutive calls to poll while consuming only from a partial set of partitions due to the fact that it clears all the records after each call. This change makes MockConsumer clearing the records only for the partitions that are not paused (whose records are actually returned by the poll). The remaining paused partitions will retain the records.

Unit test added accordingly.

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by efgpinto** [comment]
retest this please

**Comment by efgpinto** [comment]
@guozhangwang @mjsax @ableegoldman can any of you guys have a look at this MR? I don't think the test failing has anything to do with the changes. Thanks 🙏

**Comment by guozhangwang** [comment]
The proposed fix LGTM, cc @cmccabe @rajinisivaram for another review.

File: clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java]: nit: rename to `shouldNotClearRecordForPausedPartitions()`

File: clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java]: Why do we need 2 partitions for this tests?

File: clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/clients/consumer/MockConsumerTest.java]: If you want to check that records for non-pause partitions are cleared, we should have a separate test.

**Review by mjsax — APPROVED** [review]
Overall LGTM. I would simplify the test a little bit.

**Comment by efgpinto** [comment]
> Overall LGTM. I would simplify the test a little bit.

Thanks for the review and the suggestions @mjsax . Just pushed a revised version. 👍

**Comment by mjsax** [comment]
All three runs failed. Result not available any longer. Could it be a checkstyle issue?

**Comment by efgpinto** [comment]
> All three runs failed. Result not available any longer. Could it be a checkstyle issue?

Indeed it was. Wildcard import. Solved now. Tks again.

**Comment by mjsax** [comment]
Just cycling back to this PR. Seems the Jira ticket was not assigned. I found user `epinto` (not `efgpinto` as the GitHub user) -- hope that is correct?

**Comment by mjsax** [comment]
Would like to get a review from core team. \cc @cmccabe @rajinisivaram

**Comment by efgpinto** [comment]
> Just cycling back to this PR. Seems the Jira ticket was not assigned. I found user `epinto` (not `efgpinto` as the GitHub user) -- hope that is correct?

It is correct. Thanks.

**Comment by mjsax** [comment]
Ping @cmccabe and @rajinisivaram for second review and merging.

**Comment by mjsax** [comment]
Ping @hachikuji for second review and merging.

**Review by guozhangwang — APPROVED** [review]

**Pull request merged**
